core/src/main/java/com/graphhopper/routing/lm/LMRoutingAlgorithmFactory.java

```diff
@@ -0,0 +1,81 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.graphhopper.routing.lm;
+
+import com.graphhopper.routing.*;
+import com.graphhopper.routing.weighting.Weighting;
+import com.graphhopper.storage.Graph;
+import com.graphhopper.util.Parameters;
+
+public class LMRoutingAlgorithmFactory implements RoutingAlgorithmFactory {
+    private final RoutingAlgorithmFactory defaultAlgoFactory;
+    private final LandmarkStorage lms;
+    private final Weighting prepareWeighting;
+    private final int defaultActiveLandmarks;
+    private final int numBaseNodes;
+
+    public LMRoutingAlgorithmFactory(LandmarkStorage lms, int defaultActiveLandmarks, int numBaseNodes) {
+        this.defaultAlgoFactory = new RoutingAlgorithmFactorySimple();
+        this.lms = lms;
+        this.defaultActiveLandmarks = defaultActiveLandmarks;
+        this.prepareWeighting = lms.getWeighting();
+        this.numBaseNodes = numBaseNodes;
+    }
+
+    @Override
+    public RoutingAlgorithm createAlgo(Graph g, AlgorithmOptions opts) {
+        RoutingAlgorithm algo = defaultAlgoFactory.createAlgo(g, opts);
```

Comment:
Currently we create the 'default' algo with the `RoutingAlgorithmFactorySimple`, and afterwards we use `instanceof` to figure which algorithm we got and set the approximator. This is rather ugly, maybe we can simply create the algorithm in a single method (without the defaultAlgoFactory?)

Reply:
What do you mean here? If we have a single method like this - wouldn't we repeat the code from RoutingAlgorithmFactorySimple here again? (not sure if that is bad :) )

Reply:
Yes we repeat *some* of the code in `RoutingAlgorithmFactorySimple`, but really not that much and some of the code in the simple factory is 'overwritten' (for example because another approximator ist set). In my last commit I rewrote the createAlgo method without the simple factory and I like it better now. Note that its also more strict now: An error is thrown when an algorithm other than astar,astarbi,alternative_route is requested (because LM does not work with dijkstra for example). For CH its similar, it throws an error if we request non-bidir Dijkstra with CH. What do you think, does that make sense?

Reply:
Ok, yes.